typing/includemod.ml

```diff
@@ -587,7 +587,152 @@ let show_locs ppf (loc1, loc2) =
+        Printtyp.ident id context_mty rem
+  | Body x :: rem ->
+      fprintf ppf "functor (%s) ->@ %a" (argname x) context_mty rem
+  | Arg x :: rem ->
+      fprintf ppf "functor (%a : %a) -> ..." Printtyp.ident x context_mty rem
+  | [] ->
+      fprintf ppf "<here>"
+and context_mty ppf = function
+    (Module _ | Modtype _) :: _ as rem ->
+      fprintf ppf "@[<2>sig@ %a@;<1 -2>end@]" context rem
+  | cxt -> context ppf cxt
+and args ppf = function
+    Body x :: rem ->
+      fprintf ppf "(%s)%a" (argname x) args rem
+  | Arg x :: rem ->
+      fprintf ppf "(%a :@ %a) : ..." Printtyp.ident x context_mty rem
+  | cxt ->
+      fprintf ppf " :@ %a" context_mty cxt
+and argname x =
+  let s = Ident.name x in
+  if s = "*" then "" else s
+
+let alt_context ppf cxt =
+  if cxt = [] then () else
+  if List.for_all (function Module _ -> true | _ -> false) cxt then
+    fprintf ppf "in module %a,@ " Printtyp.path (path_of_context cxt)
+  else
+    fprintf ppf "@[<hv 2>at position@ %a,@]@ " context cxt
+
+let context ppf cxt =
+  if cxt = [] then () else
+  if List.for_all (function Module _ -> true | _ -> false) cxt then
+    fprintf ppf "In module %a:@ " Printtyp.path (path_of_context cxt)
+  else
+    fprintf ppf "@[<hv 2>At position@ %a@]@ " context cxt
+
+
+module Illegal_permutation = struct
+  (** Extraction of information in case of illegal permutation
+      in a module type *)
+
+  (** When examining coercions, we only have runtime component indices,
+      we use thus a limited version of {!pos}. *)
+  type coerce_pos =
+    | Item of int
+    | InArg
+    | InBody
+
+  let either f x g y = match f x with
+    | None -> g y
+    | Some _ as x -> x
+
+  (** We extract a lone transposition from a full tree of permutations. *)
+  let rec transposition path = function
+    | Tcoerce_structure(c,_) ->
+        either
+          (not_fixpoint path 0) c
+          (first_non_id path 0) c
+    | Tcoerce_functor(arg,res) ->
+        either
+          (transposition (InArg::path)) arg
+          (transposition (InBody::path)) res
+    | Tcoerce_none -> None
+    | Tcoerce_alias _ | Tcoerce_primitive _ ->
+        (* these coercions are not inversible, and raise an error earlier when
+           checking for module type equivalence *)
+        assert false
+  (* we search the first point which is not invariant at the current level *)
+  and not_fixpoint path pos = function
+    | [] -> None
+    | (n, _) :: q when n = pos -> not_fixpoint path (pos+1) q
+    | (n,_) :: _ -> Some(List.rev path, pos, n)
+  (* we search the first item with a non-identity inner coercion *)
+  and first_non_id path pos = function
+    | [] -> None
+    | (_,Tcoerce_none) :: q -> first_non_id path (pos + 1) q
+    | (_,c) :: q ->
+        either
+          (transposition (Item pos :: path)) c
+          (first_non_id path (pos + 1)) q
+
+  let transposition c = match transposition [] c with
+    | None -> raise Not_found
+    | Some x -> x
```

Comment:
Nitpicking: this style of shadowing makes thing more painful to read in absence of solid jump-to-definition features. What about calling the first function `transposition_under`?

Reply:
I'm not sure I agree with this -- functions for internal use of other functions I think are best shadowed or made local.  Maybe it could be made local in this case?  (Also, please put `match` on the next line...)

Reply:
Having the helper function local to the outside function is fine with me (they can have the same name, except when the outer function is also recursive, because this is unpalatable shadowing again).

That said, in my own experience the worker function is sometimes useful to access from other parts of the code, because it actually exposes a more "modular" interface than the wrapper (this is the programming equivalent of "the lemma with a strengthened induction hypothesis"). This is actually the case here, as `transposition_under` is also used in `first_non_id`. I think this is the code trying to tell us that the worker function deserves its own name.